docs/how-to/gpu-enabled-mpi.rst

```diff
@@ -0,0 +1,260 @@
+GPU-enabled MPI
+***************************************************************************************************
+
+The Message Passing Interface (`MPI <https://www.mpi-forum.org>`_) is a standard API for distributed
+and parallel application development that can scale to multi-node clusters. To facilitate the porting of
+applications to clusters with GPUs, ROCm enables various technologies. You can use these
+technologies add GPU pointers to MPI calls and enable ROCm-aware MPI libraries to deliver optimal
+performance for both intra-node and inter-node GPU-to-GPU communication.
+
+The AMD kernel driver exposes remote direct memory access (RDMA) through *PeerDirect* interfaces.
+This allows network interface cards (NICs) to directly read and write to RDMA-capable GPU device
+memory, resulting in high-speed direct memory access (DMA) transfers between GPU and NIC. These
+interfaces are used to optimize inter-node MPI message communication.
+
+The Open MPI project is an open source implementation of the MPI. It's developed and maintained by
+a consortium of academic, research, and industry partners. To compile Open MPI with ROCm support,
+refer to the following sections:
+
+* :ref:`open-mpi-ucx`
+* :ref:`open-mpi-libfabric`
+
+.. _open-mpi-ucx:
+
+ROCm-aware Open MPI on InfiniBand and RoCE networks using UCX
+================================================================
+
+The `Unified Communication Framework <https://www.openucx.org/documentation>`_ (UCX), is an
+open source, cross-platform framework designed to provide a common set of communication
+interfaces for various network programming models and interfaces. UCX uses ROCm technologies to
+implement various network operation primitives. UCX is the standard communication library for
+InfiniBand and RDMA over Converged Ethernet (RoCE) network interconnect. To optimize data
+transfer operations, many MPI libraries, including Open MPI, can leverage UCX internally.
+
+To set up UCX so that it can be used to compile Open MPI, refer to the following instructions.
+
+1. Set environment variables to install all software components in the same base directory. We use the
+home directory in our example, but you can specify a different location if you want.
+
+    .. code-block:: shell
+
+        export INSTALL_DIR=$HOME/ompi_for_gpu
+        export BUILD_DIR=/tmp/ompi_for_gpu_build
+        mkdir -p $BUILD_DIR
+
+2. Instal UCX. To view UCX and ROCm version compatibility, refer to the
+`communication libraries tables <https://rocm.docs.amd.com/en/latest/reference/gpu_libraries/communication.html>`_.
+
+    .. code-block:: shell
+
+        export UCX_DIR=$INSTALL_DIR/ucx
+        cd $BUILD_DIR
+        git clone https://github.com/openucx/ucx.git -b v1.15.x
+        cd ucx
+        ./autogen.sh
+        mkdir build
+        cd build
+        ../configure -prefix=$UCX_DIR \
+            --with-rocm=/opt/rocm
+        make -j $(nproc)
+        make -j $(nproc) install
+
+3. Install Open MPI.
```

Comment:
@samjwu do you know what the alignment is wonky with #3 (in the html version)?

Reply:
RST seems to be indenting 3. due to link in 2.

https://snippets.documatt.com/

Reply:
Any chance you know of a way to fix it without removing that link?

Reply:
Can you use ssh based clones instead? Figure out the details later. I think it is linkify messing things up.

Reply:
the https version is usually preferred in such a scenario since it doesn't require a github account to perform the checkout. However, we can make the change for now to make sure that this gets in 6.0

The git clone in line 53 would have to be replaced by
```
git clone git@github.com:openucx/ucx.git -b v1.15.x
```

in lines 69/70  and lines 237/238 with
```
git clone --recursive  git@github.com:open-mpi/ompi.git -b v5.0.x
```

in line 161
```
git clone git@github.com:openucx/ucc.git -b v1.2.x
```

in line 220
```
git clone git@github.com:ofiwg/libfabric.git -b v1.19.x
```

Reply:
Maybe the correct solution long term is to use the released tar files of the libraries instead of using the git repositories (i.e. `wget` instead of `git clone`), but since this changes the required steps for the compilation (e.g. no autogen required) I would prefer to do that in a later stage